rascal-vscode-extension/src/lsp/JsonOutputChannel.ts

```diff
@@ -0,0 +1,190 @@
+                break;
+            }
+            default: {
+                this.logChannel.appendLine(`${loglevel} ${message}`);
+            }
+        }
+    }
+
+    private printPayload(payload: string): void {
+        for (const line of payload.trim().split("\n")) {
+            try {
+                const res = this.printJsonPayLoad(line);
+                if (res === undefined) {
+                    // it's JSON, but we don't understand the shape
+                    this.logChannel.appendLine(line);
+                }
+            } catch (e) {
+                if (e instanceof SyntaxError) {
+                    // this was not JSON at all
+                    this.printNonJsonPayload(line);
+                } else {
+                    this.logChannel.appendLine(`Error while logging ${line}: ${e}`);
+                    throw e;
+                }
+            }
+        }
+    }
+
+    private formatServerTime(date: Date | string): string {
+        if (typeof date === "string") {
+            date = new Date(date);
+        }
+        return date.toISOString();
+    }
+
+    private formatMessage(thread: string, originalTime: Date | string, message: string, loggerName?: string) {
+        const loggerPart = loggerName ? ` ${loggerName}` : "";
+        return `[${thread}] (${this.formatServerTime(originalTime)})${loggerPart} ${message}`;
+    }
+
+    private printNonJsonPayload(payload: string): void {
+        // try to find the log level and message
+        const log4jDefault = /^\s*(?<timestamp>[^\s]+)\s+(?<thread>\w+)\s+(?<level>\w+)\s+(?<message>.*)$/si;
+
+        const matches = log4jDefault.exec(payload);
+        if (matches?.groups
+            && "thread" in matches.groups
+            && "level" in matches.groups
+            && "message" in matches.groups
+            && "timestamp" in matches.groups) {
+            const logLevel = matches.groups["level"].toLocaleUpperCase() as LogLevel;
+            this.printLogOutput(logLevel, this.formatMessage(matches.groups["thread"], matches.groups["timestamp"], matches.groups["message"]));
+        } else {
+            // we do not understand the structure of this log message; just print it as-is
+            this.logChannel.appendLine(payload);
+        }
+    }
+
+    private printJsonPayLoad(payload: string): LogMessage | undefined {
+        const json = JSON.parse(payload);
+        const log = LogMessage.cast(json);
+        if (log) {
+            // no timestamp or log level, since LogOutputChannel functions add those
+            this.printLogOutput(log.level, this.formatMessage(log.threadName, log.timestamp, log.message, log.loggerName));
+            return log;
+        }
+        return undefined;
+    }
+
+    append(payload: string): void {
+        this.printPayload(payload);
+    }
+
+    appendLine(payload: string): void {
+        this.printPayload(payload);
+    }
+
+    show(preserveFocus?: unknown): void {
```

Comment:
what if we pass a string? nothing happens? this seems a bit of a weird type? shouldn't it be `show(preserverFocus: boolean | undefined)`?

Reply:
you can have multiple signatures for a function I think, i forgot the exact syntax.

```typescript
show(preserveFocus?: boolean): void
show(column?: ViewColumn, preserveFocus?: boolean): void {
}
```

but I agree, kinda weird these 2 overloads.